symengine/matrices/matrix_add.cpp

```diff
@@ -0,0 +1,176 @@
+            expanded.insert(expanded.end(), container.begin(), container.end());
+        } else {
+            expanded.push_back(term);
+        }
+    }
+    check_matching_sizes(expanded);
+    vec_basic keep;
+    RCP<const DiagonalMatrix> diag;
+    RCP<const ImmutableDenseMatrix> dense;
+    RCP<const ZeroMatrix> zero;
+    for (auto &term : expanded) {
+        if (is_a<ZeroMatrix>(*term)) {
+            zero = rcp_static_cast<const ZeroMatrix>(term);
+        } else if (is_a<DiagonalMatrix>(*term)) {
+            if (diag.is_null()) {
+                diag = rcp_static_cast<const DiagonalMatrix>(term);
+            } else {
+                vec_basic container;
+                for (size_t i = 0; i < diag->get_container().size(); i++) {
+                    container.push_back(
+                        add(diag->get_container()[i],
+                            down_cast<const DiagonalMatrix &>(*term)
+                                .get_container()[i]));
+                }
+                diag = make_rcp<const DiagonalMatrix>(container);
+            }
+        } else if (is_a<ImmutableDenseMatrix>(*term)) {
+            if (dense.is_null()) {
+                dense = rcp_static_cast<const ImmutableDenseMatrix>(term);
+            } else {
+                const vec_basic &vec1
+                    = down_cast<const ImmutableDenseMatrix &>(*term)
+                          .get_values();
+                const vec_basic &vec2 = dense->get_values();
+                vec_basic sum(vec1.size());
+                for (size_t i = 0; i < vec1.size(); i++) {
+                    sum[i] = add(vec1[i], vec2[i]);
+                }
+                dense = make_rcp<const ImmutableDenseMatrix>(
+                    dense->nrows(), dense->ncols(), sum);
+            }
+        } else {
+            keep.push_back(term);
+        }
+    }
+    if (!diag.is_null()) {
+        if (!dense.is_null()) {
+            // Add diagonal with dense matrix
+            auto vec = dense->get_values();
+            vec_basic sum;
+            for (size_t i = 0; i < dense->nrows(); i++) {
+                for (size_t j = 0; j < dense->ncols(); j++) {
+                    if (i == j) {
+                        sum.push_back(add(dense->get(i, j), diag->get(i)));
+                    } else {
+                        sum.push_back(dense->get(i, j));
+                    }
+                }
+            }
+            dense = make_rcp<const ImmutableDenseMatrix>(dense->nrows(),
+                                                         dense->ncols(), sum);
+        } else {
+            keep.push_back(diag);
+        }
+    }
+    if (!dense.is_null()) {
+        keep.push_back(dense);
+    }
+    if (keep.size() == 1) {
+        return rcp_static_cast<const MatrixExpr>(keep[0]);
+    }
+    if (keep.size() == 0 && !zero.is_null()) {
+        return zero;
+    }
+    return make_rcp<const MatrixAdd>(keep);
```

Comment:
`I + I` should return `2*I`. In order to have this representation, should we have a dictionary instead of a vector?

Reply:
`MatrixMul` has a `scalar` element for this purpose.

Reply:
How does `2*X + 3*X` work?

Reply:
I would think `MatrixAdd(MatrixMul(2, X), MatrixMul(3, X))`. Not sure if this works currently though.

Reply:
This is indeed how this works. I added an extra unit test for this example and also fixed a bug in `get_args` for `MatrixMul`